torch/csrc/jit/python/python_list.h

```diff
@@ -0,0 +1,233 @@
+#include <pybind11/detail/common.h>
+#include <torch/csrc/utils/pybind.h>
+#include <cstddef>
+#include <stdexcept>
+
+namespace torch {
+namespace jit {
+
+void initScriptListBindings(PyObject* module);
+
+/// An iterator over the elements of ScriptList. This is used to support
+/// __iter__(), .
+class ScriptListIterator final {
+ public:
+  ScriptListIterator(
+      c10::impl::GenericList::iterator iter,
+      c10::impl::GenericList::iterator end)
+      : iter_(iter), end_(end) {}
+  IValue next();
+  bool done() const;
+
+ private:
+  c10::impl::GenericList::iterator iter_;
+  c10::impl::GenericList::iterator end_;
+};
+
+/// A wrapper around c10::List that can be exposed in Python via pybind
+/// with an API identical to the Python list class. This allows
+/// lists to have reference semantics across the Python/TorchScript
+/// boundary.
+class ScriptList final {
+ public:
+  // TODO: Do these make sense?
+  using size_type = size_t;
+  using diff_type = ssize_t;
+
+  // Constructor for empty lists created during slicing, extending, etc.
+  ScriptList(const TypePtr& type) {
+    auto list_type = type->expect<ListType>();
+    auto d = c10::impl::GenericList(list_type);
+    list_ = IValue(d);
+  }
+
+  // Constructor for instances based on existing lists (e.g. a
+  // Python instance or a list nested inside another).
+  ScriptList(IValue data) {
+    TORCH_INTERNAL_ASSERT(data.isList());
+    list_ = std::move(data);
+  }
+
+  ListTypePtr type() const {
+    return list_.type()->cast<ListType>();
+  }
+
+  // Return a string representation that can be used
+  // to reconstruct the instance.
+  std::string repr() const {
+    std::ostringstream s;
+    s << '[';
+    bool f = false;
+    for (auto const& elem : list_.toList()) {
+      if (f) {
+        s << ", ";
+      }
+      s << IValue(elem);
+      f = true;
+    }
+    s << ']';
+    return s.str();
+  }
+
+  // Return an iterator over the elements of the list.
+  ScriptListIterator iter() const {
+    auto begin = list_.toList().begin();
+    auto end = list_.toList().end();
+    return ScriptListIterator(begin, end);
+  }
+
+  // Interpret the list as a boolean; empty means false, non-empty means
+  // true.
+  bool toBool() const {
+    return !(list_.toList().empty());
+  }
+
+  // Get the value for the given index.
+  IValue getItem(diff_type idx) {
+    idx = wrap_index(idx);
+    return list_.toList().get(idx);
+  };
+
+  // Set the value corresponding to the given index.
+  // TODO: Handle negative and wraparound.
+  void setItem(diff_type idx, const IValue& value) {
+    idx = wrap_index(idx);
+    return list_.toList().set(idx, value);
+  }
+
+  // Check whether the list contains the given value.
+  bool contains(const IValue& value) {
+    for (const auto& elem : list_.toList()) {
+      if (elem == value) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  // Delete the item at the given index from the list.
+  void delItem(size_type idx) {
```

Comment:
Why isn't signed `diff_type` allowed here like in `setItem`?

Reply:
Good call